### DISCUSSION TITLE
Problem trying to convert (const char*) type string with Japanese characters to std::wstringstream

### DISCUSSION THREAD

**Discussion by IgnacioBMY** [fffaraz/awesome-cpp#1255] [Q&A]
Hello all,

Thanks for let me write here.

I have a code like this:

_//xmlDocument is a variable of type tinyxml2::XMLDocument and it contains Japanese characters_
tinyxml2::XMLPrinter var_xmlp(nullptr, true, 0);
xmlDocument.Print(&var_xmlp);
auto result = var_xmlp.CStr();
std::wstringstream stream;
stream << result;

While I am debugging I can see that "result" contains the correct Japanese characters; however, after assign "result" to "stream", the Japanese characters are garbled into "stream".

Also, due to my company's requirements, I have to solve this issue using standard library.

So, my question is: does anyone know a way to convert a (const char*) string with Japanese characters to a std::wstringstream string using just the standard library (STL)?

Thank you very much for your help and best regards,

Ignacio

**Comment by Josh-Ruben** [comment]
The issue is an encoding conversion, not simply a conversion from const char* to wchar_t.

tinyxml2::XMLPrinter::CStr() returns a const char*, and TinyXML-2 uses UTF-8 for XML. Therefore, result contains UTF-8 encoded bytes.

When you do:

std::wstringstream stream;
stream << result;

the bytes are not necessarily interpreted as UTF-8 and decoded into Unicode characters. You need to explicitly perform the UTF-8 conversion before inserting the result into the wide stream.

For C++11/C++14/C++17, assuming the XML is UTF-8:

#include <codecvt>
#include <locale>
#include <sstream>
#include <string>

tinyxml2::XMLPrinter var_xmlp(nullptr, true, 0);
xmlDocument.Print(&var_xmlp);

const char* result = var_xmlp.CStr();

std::wstring_convert<std::codecvt_utf8<wchar_t>> converter;
std::wstring wideResult = converter.from_bytes(result);

std::wstringstream stream;
stream << wideResult;

Or, if you ultimately only need a std::wstring, the std::wstringstream may not be necessary:

std::wstring_convert<std::codecvt_utf8<wchar_t>> converter;
std::wstring wideResult = converter.from_bytes(var_xmlp.CStr());

The important distinction is that a const char* is a sequence of bytes; the encoding must be known before converting it to a wide string.

In this case, assuming the TinyXML-2 output is UTF-8, std::codecvt_utf8<wchar_t> performs the required UTF-8-to-wide-character conversion.

Note: <codecvt> and std::wstring_convert were deprecated in C++17 and later removed from the standard library. However, for code targeting the C++11–C++17 era of this question, this is a standard-library solution.
